docs/how-to/define-state-machines.md

```diff
@@ -0,0 +1,427 @@
+
+---
+
+## When to Use State Machines
+
+State machines are useful when your component has distinct **modes** or **operational states** with different behavior in specific states and rules for the transitions between states. Examples include:
+
+* A radio with `OFF`, `IDLE`, and `TRANSMITTING` states.
+* A sensor with a series of start-up states
+
+By modeling these as a state machine, you make system behavior explicit, verifiable, and easier to test.
+
+---
+
+## Example State Machine
+
+This guide will walk through the steps of implementing a state machine with the following properties:
+
+0. The initial state is `RESET`. Followed by `WAIT_RESET`, `ENABLE`, `CONFIGURE` and `RUN`.
+1. At each state a `tick` signal will trigger an action.
+2. Actions may yield `success` or `error` signals.
+3. `success` signals progress linearly through to the next state.
+4. `error` signals return to state `RESET`
+5. The state machine remains in a given state (and repeats the action) when the action yields no signals.
+
+This state machine can be seen in the following diagram.
+
+```mermaid
+stateDiagram-v2
+    [*] --> RESET
+
+    %% State definitions with annotations
+    state "RESET
+
+    tick / doReset" as RESET
+
+    state "WAIT_RESET
+
+    tick / checkReset" as WAIT_RESET
+
+    state "ENABLE
+
+    tick / doEnable" as ENABLE
+
+    state "CONFIGURE
+
+    tick / doConfigure" as CONFIGURE
+
+    state "RUN
+
+    tick / doRead" as RUN
+
+    RESET --> WAIT_RESET : success
+    WAIT_RESET --> ENABLE : success
+    ENABLE --> CONFIGURE : success
+    CONFIGURE --> RUN : success
+
+    ENABLE --> RESET : error
+    WAIT_RESET --> RESET : error
+    CONFIGURE --> RESET : error
+    RUN --> RESET : error
+```
+
+This state machine is implemented in the [MpuImu](TODO) component. This component drives the "next" signal from a rate group handler.
+
+> [!NOTE]
+> The [MpuImu](TODO) component implements a further `reconfigure` signal allowing the machine to return to reconfigure state.
+
+## Designing a State Machine in FPP
+
+We will model the Imu state machine in FPP.  Typically users would define state machines in a new file.  We should use `ImuStateMachine.fpp` defined in the same folder (i.e. module) as the `ImuManager` this will support.  Define the basic module and state machine with:
+
+```
+module MpuImu {
+    state machine ImuStateMachine
+}
+```
+
+The module, filename, etc was chosen to remain consistent with our example code (and the `ImuManager` component that uses the state machine).
+
+
+### Defining a State Machine and Initial State
+
+The first step of defining our state machine, is to name the machine and the initial `RESET` state. This step will give us a basis for state machine implementation.
+
+```
+module MpuImu {
+    @ Define ImuStateMachine State Machine
+    state machine ImuStateMachine {
+        @ Initial state: reset the device
+        initial enter RESET
+
+        @ Reset the Imu
+        state RESET {
```

Comment:
You may want to omit the braces `{}` here and add them later, when they are needed. Otherwise it may create the impression that braces are required in all cases, even when they are empty.